ksm.go

```diff
@@ -0,0 +1,303 @@
+	ksmStandard                  = "standard"
+	ksmAggressive                = "aggressive"
+)
+
+func (k ksmSettingKnob) String() string {
+	switch k {
+	case ksmSlug:
+		return "slug"
+	case ksmSlow:
+		return "slow"
+	case ksmStandard:
+		return "standard"
+	case ksmAggressive:
+		return "aggressive"
+	}
+
+	return ""
+}
+
+func (k *ksmSettingKnob) Set(value string) error {
+	fmt.Printf("value %s\n", value)
+	for _, r := range strings.Split(value, ",") {
+		if r == "slug" {
+			*k = ksmSlug
+			return nil
+		} else if r == "slow" {
+			*k = ksmSlow
+			return nil
+		} else if r == "standard" {
+			*k = ksmStandard
+			return nil
+		} else if r == "aggressive" {
+			*k = ksmAggressive
+			return nil
+		} else {
+			return errors.New("Unknown KSM knob")
+		}
+	}
+
+	return nil
+}
+
+var ksmSettings = map[ksmSettingKnob]ksmSetting{
+	ksmSlug:       {1000, 500}, // Every 500ms, we scan 1 page for every 1000 pages available in the system
+	ksmSlow:       {500, 100},  // Every 100ms, we scan 1 page for every 500 pages available in the system
+	ksmStandard:   {100, 10},   // Every 10ms, we scan 1 page for every 100 pages available in the system
+	ksmAggressive: {10, 1},     // Every ms, we scan 1 page for every 10 pages available in the system
+}
+
+const (
+	ksmRoot           = "/sys/kernel/mm/ksm/"
+	ksmRunFile        = "run"
+	ksmPagesToScan    = "pages_to_scan"
+	ksmSleepMillisec  = "sleep_millisecs"
+	ksmRun            = "1"
+	defaultKSMSetting = ksmAggressive
+)
+
+type sysfsAttribute struct {
+	path string
+	file *os.File
+}
+
+func (attr *sysfsAttribute) open() error {
+	file, err := os.OpenFile(attr.path, os.O_RDWR|syscall.O_NONBLOCK, 0660)
+	attr.file = file
+	return err
+}
+
+func (attr *sysfsAttribute) close() error {
+	err := attr.file.Close()
+	attr.file = nil
+
+	return err
+}
+
+func (attr *sysfsAttribute) read() (string, error) {
+	_, err := attr.file.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return "", err
+	}
+
+	data, err := ioutil.ReadAll(attr.file)
+	if err != nil {
+		return "", err
+	}
+
+	return string(data), nil
+}
+
+func (attr *sysfsAttribute) write(value string) error {
+	_, err := attr.file.Seek(0, os.SEEK_SET)
+	if err != nil {
+		return err
+	}
+	_, err = attr.file.WriteString(value)
+
+	return err
+}
+
+// store a sysfs attribute and returns its old value
+func (attr *sysfsAttribute) store(value string) (string, error) {
+	if attr.file == nil {
+		err := attr.open()
+		if err != nil {
+			return "", err
+		}
+
+		defer func() {
+			_ = attr.close()
+		}()
+	}
+
+	oldValue, err := attr.read()
+	if err != nil {
+		return "", err
+	}
+
+	err = attr.write(value)
+	if err != nil {
+		return "", err
+	}
+
+	return oldValue, nil
+}
+
+func isKSMAvailable(ksmDir string) error {
+	info, err := os.Stat(ksmDir)
+	if err != nil || !info.IsDir() {
+		return fmt.Errorf("%s is not available", ksmDir)
+	}
+
+	return nil
+}
+
+func systemPages() (int64, error) {
+	// We're going to parse meminfo
+	f, err := os.Open("/proc/meminfo")
```

Comment:
Again, it would be good if this was defined as a global `var` to allow the tests to modify it.